File: day04/main.go
package main

import (
	aoc "aoc/utils"
	"fmt"
	"log"
	"math"
	"strconv"
	"strings"
)

type scratchcard struct {
	number         int
	winningNumbers []int
	numbers        []int
}

func getPoints(card scratchcard) int {
	// fmt.Println(card)
	var wins int
	for _, number := range card.numbers {
		for _, winningNumber := range card.winningNumbers {
			if number == winningNumber {
				wins += 1
			}
		}
	}
	// fmt.Println("wins: ", wins)
	return int(math.Exp2(float64(wins - 1)))
}

func main() {
	lines, err := aoc.ReadLines("./input.txt")
	if err != nil {
		log.Fatal(err)
	}

	// fmt.Println(lines)

	var scratchcards []scratchcard
	for i, line := range lines {
		card := scratchcard{number: i + 1}

		s := strings.Split(line, ": ")[1]
		bothNumers := strings.Split(s, " | ")

		for _, number_s := range strings.Split(bothNumers[0], " ") {
			if number_s == "" {
				continue
			}
			number, err := strconv.Atoi(number_s)
			if err != nil {
				log.Fatal(err)
			}
			card.winningNumbers = append(card.winningNumbers, number)
		}

		for _, number_s := range strings.Split(bothNumers[1], " ") {
			if number_s == "" {
				continue
			}
			number, err := strconv.Atoi(number_s)
			if err != nil {
				log.Fatal(err)
			}
			card.numbers = append(card.numbers, number)
		}

		scratchcards = append(scratchcards, card)
	}

	// fmt.Println(scratchcards)

	var pointsSum int
	for _, card := range scratchcards {
		points := getPoints(card)
		// fmt.Println(points)
		pointsSum += points
	}
	fmt.Println(pointsSum)
}

File: day04/main.py
import re

with open("input.txt", "r") as file:
    lines = [l.strip() for l in file.readlines()]


pattern = r"\d+"
r = re.compile(pattern)

card_counter = [1] * len(lines)
for i, line in enumerate(lines):
    numbers_str = line.split(":")[1].split("|")

    winning_numbers = r.findall(numbers_str[0])
    numbers = r.findall(numbers_str[1])

    intersection = set(numbers).intersection(winning_numbers)

    for j in range(len(intersection)):
        # for each of the next #wins cards
        # add 1 count for each existing current card
        card_counter[i + j + 1] += card_counter[i]

total_cards = sum(card_counter)
print("Task 2:", total_cards)
